### PR TITLE
Include transfer learning example for backtesting

### PR DESCRIPTION
**Pull request by AVHopp**
This PR introduces two basic transfer learning examples.

The first example is a backtesting example that just shows how to initialize and use a `TaskParamter`. In particular, things like `active_values` and more detailed explanations are NOT INTENDED to be part of that example.
Instead, a second example (which is NOT YET included here) will be constructed. This example will guide the user through a more detailed application of the current transfer learning capabilities, including the use of `active_values` and more meaningful lookup tables.
The reason for this is that it is not perfectly clear yet which aspects of the corresponding examples should stay in the examples and which should rather be part of the user guide which is something we will need to discuss.

I intentionally already provided this example independent of the current refactorings regarding the simulation. We might want to finish them first and then include this example or the other way round.

### CODE REVIEW

File: examples/Backtesting/transfer_learning.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+# We can now use the `simulate_transfer_learning` function to simulate a full experiment.
+
+results = simulate_transfer_learning(
+    campaign,
+    lookup,
+    batch_quantity=1,
+    n_doe_iterations=N_DOE_ITERATIONS,
+    n_mc_iterations=N_MC_ITERATIONS,
+)
+
```
Comment (Scienfitz) [inline examples/Backtesting/transfer_learning.py:116]: theres no comparison to just plain simulating the campaigns without transfer learning which feels kind of esssential for such a demo?

File: examples/Backtesting/transfer_learning.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+# The following lines plot the results and save the plot in run_analytical.png
+sns.lineplot(data=results, x="Num_Experiments", y="Target_CumBest", hue="Scenario")
+plt.gcf().set_size_inches(24, 8)
+plt.savefig("./run_analytical.png")
```
Comment (Scienfitz) [inline examples/Backtesting/transfer_learning.py:120]: change name to `run_transfer_learning` or similar

**Review by Scienfitz — CHANGES_REQUESTED** [review]

File: examples/Backtesting/transfer_learning.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
+### Example for a simulation loop using the transfer learning capabilities
```
Comment (Scienfitz) [inline examples/Backtesting/transfer_learning.py:1]: after some more thought here a proposal for the overall content of the transfer learnign examples:

1) The actual demo of how it would be used in practice - not backtesting
- generate searchspae with a taskparaemter that has `values = ['old_campaign1','old_campaign2','current_campaign']` and `active_values = ['current campaign']` (or better examples closer to the real world)
- add the existsing data with `add_measurements`
- now performa a small recommendation loop and demosntrate that we get recommendations for `current_campaign`


2) A backteting example
- basically whats in here
- but also demonstrating the improvement the result have when compared to simualting without transfer learning (ie 2 additional curves)

**Review by Scienfitz — CHANGES_REQUESTED** [review]

File: examples/Backtesting/transfer_learning.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+# Generate the grid coordinates using numpy
+x = parameters[0].values
+y = parameters[1].values
+xx, yy = np.meshgrid(x, y)
+
+lookup = pd.DataFrame({"x_1": xx.flatten(), "x_2": yy.flatten()})
+
+# We randomly decide for each row which function we measure
+lookup["Function"] = np.random.choice(["ThreeHump", "SixHump"], lookup.shape[0])
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:85]: That was due to me not fully understanding the logic. This will be changed in the upcoming version.

File: examples/Backtesting/transfer_learning.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+x1 = NumericalDiscreteParameter(
+    name="x_1", values=list(np.linspace(BOUNDS[0, 0], BOUNDS[1, 0], POINTS_PER_DIM))
+)
+x2 = NumericalDiscreteParameter(
+    name="x_2", values=list(np.linspace(BOUNDS[0, 1], BOUNDS[1, 1], POINTS_PER_DIM))
+)
+
+# We now append a ``TaskParameter``. This parameter contains the information about the
+# different tasks that we have in this transfer learning example.
+task_param = TaskParameter(name="Function", values=("ThreeHump", "SixHump"))
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:60]: As agreed upon, this example has been changed into a backtesting example that will not include this feature, hence this conversation is being resolved.

File: examples/Backtesting/transfer_learning.py (hunk 1 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:106]: Such a comparison is included now. This is the output with the current configuration:
image
It is the best I could come up with while keeping the runtime reasonable. If this is acceptable, please resolve.

File: examples/Backtesting/transfer_learning.py (hunk 3 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:1]: The current file is what you propose as 2nd file here. I will not resolve this comment until I have implemented a variant of the first demo you mention here.

File: examples/Backtesting/transfer_learning.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+campaign_without_tl_3hc = Campaign(
+    searchspace=SearchSpace.from_product([x1, x2]), objective=objective
+)
+campaign_without_tl_6hc = Campaign(
+    searchspace=SearchSpace.from_product([x1, x2]), objective=objective
+)
+
+scenarios_3hc = {"ThreeHump no TL": campaign_without_tl_3hc}
+scenarios_6hc = {"SixHump no TL": campaign_without_tl_6hc}
+
```
Comment (Scienfitz) [inline examples/Backtesting/transfer_learning.py:120]: could be simplified by making a single dict and a single call to `simulate_scenarios`

**Review by Scienfitz — CHANGES_REQUESTED** [review]

File: examples/Backtesting/transfer_learning.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
+### Example for a simulation loop using the transfer learning capabilities
```
Comment (Scienfitz) [inline examples/Backtesting/transfer_learning.py:1]: the new example should be mentioned in the changelog

**Review by Scienfitz — CHANGES_REQUESTED** [review]

File: examples/Backtesting/transfer_learning.py (hunk 1 above)
Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:106]: not sure why youre so stingy on the MC runs, this takes maybe 45min (and n iteraions cans till be reduced):
image

this actually doesnt look that great for transfer learning, are we sure everything is correct?

Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:106]: I wanted to keep number of MC iterations low since examples are tested/executed in our tox pipelines if you do not configure them to be ignored. So I wanted to keep the overall execution time below 1min.
I do not know. Maybe the example is not very well suited for a Transfer Learning example?

File: examples/Backtesting/transfer_learning.py (hunk 6 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:120]: Are you sure? They need different lookup functions, hence different calls to `simulate_scenarios`

Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:120]: they can share a lookup as long as the campaigns are configured correctly (which then would indeed pose a usecase for `active_values`)

File: examples/Backtesting/transfer_learning.py (hunk 1 above)
Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:106]: I was not talking about increasing the number of mc runs in the published example, but just for us to check. did you not want to check whether this example actually makes sense and produces a reasonable result? it can reamin with n_mc=2 for the published version

Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:106]: i thought these functions were specifically designed for a transfer learning use case?

File: examples/Backtesting/transfer_learning.py (hunk 6 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:120]: Ok, but then there is no need for having a dedicated backtesting and a dedicated active_values example, right? Then these would probably collapse into one example

File: examples/Backtesting/transfer_learning.py (hunk 1 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:106]: Well, I had a look at the available test functions in botorch and these were the ones that seemed to make the most sense for me, given that I wanted to have a low-dimensional problem. I will however test it again with some higher dimensional variations of the Hartmann function - there is an unconstrained version as well as two constrained versions. This might also be a nice use case for active values: After having learned the unconstrained function and a constrained variant, learn the other constrained variant.

File: examples/Backtesting/transfer_learning.py (hunk 6 above)
Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:120]: its still mentioned in the userguide so no need to restrict this example here

File: examples/Backtesting/transfer_learning.py (hunk 1 above)
Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:106]: not sure thats the right approach, the transfer learning will only yield better results if the two examples chosen have some common behavior. Is that in any way indicated for the hump functions used there or the hartmannX functions?
might be required to make some custom examples (maybe get inspiration form some neurips papers, there might be some that used analytical examples, otherwise... we also have the public reaction data from the benchmark)

File: examples/Backtesting/transfer_learning.py (hunk 6 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:120]: But `simulate_transfer_learning` only accepts a single campaign, and not a dict, so this would require an outer loop in that case, is this better?

File: examples/Backtesting/transfer_learning.py (hunk 1 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:106]: These produce quite nice plots in a reasonable time:
image
So I'd propose to use these for the example here.

File: examples/Backtesting/transfer_learning.py (hunk 6 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:134]: I have tried to make the code better to read and to understand by using a `for` loop, since I did not manage to make it run using any of the simulation functions.

File: examples/Backtesting/transfer_learning.py (hunk 1 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:129]: I have changed the example accordingly, this is now the latest plot:
image
Please resolve if this is gucci.

Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:129]: 1) did you make changes to the botorch version? Locally on your branch Im getting import error indicating there have been changes in botorch verisons regarding these functions (im on 0.8.5)

2) Please reduce the number of doe and mc iterations to the numbers used in other examples

File: examples/Backtesting/transfer_learning.py (hunk 3 above)
Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:1]: ? that should be straighforward

- Tl parameter should not be excluded from searchspace but instead have `active_values='name_of_the_particular_function'`
- This will result in 3 independen campaign objects
- These can be simulated with `simulate_scenarios`
- If you use a dataframe as lookup then they can share one single dataframe as lookup

Id say its not really acceptable to code your own look as this example is about our backtest utilities

File: examples/Backtesting/transfer_learning.py (hunk 6 above)
Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:134]: ? that should be straighforward

- Tl parameter should not be excluded from searchspace but instead have `active_values='name_of_the_particular_function'`
- This will result in 3 independen campaign objects
- These can be simulated with `simulate_scenarios`
- If you use a dataframe as lookup then they can share one single dataframe as lookup

Id say its not really acceptable to code your own look as this example is about our backtest utilities

File: examples/Backtesting/transfer_learning.py (hunk 1 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:129]: I had to re-install my environment, then it worked.

Comment (Scienfitz) [reply examples/Backtesting/transfer_learning.py:129]: so with what botorch versions does this work? it needs to be correct in pyproject.toml

Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:129]: According to changelog, it's 0.9.3. Will add.

File: examples/Backtesting/transfer_learning.py (hunk 6 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py]: This was now changed, please resolve if the current variant works.

File: examples/Backtesting/transfer_learning.py (hunk 1 above)
Comment (AVHopp) [reply examples/Backtesting/transfer_learning.py:132]: Since we did already discuss this issue, I'll resolve here and create an additional comment to avoid accidental merging.

File: pyproject.toml [hunk 8]
```diff
@@ -30,7 +30,7 @@ keywords = [
 dynamic = ['version']
 dependencies = [
     "attrs>=22.2.0",
-    "botorch>=0.8.1",
+    "botorch>=0.9.3",
```
Comment (AVHopp) [inline pyproject.toml:33]: For this reason, we need to wait until we dropped python 3.8. This comment is here in order to block merging.

**Review by AdrianSosic — CHANGES_REQUESTED** [review]
Hi @AVHopp, thanks for drafting the example. In this first step, I don't want to give detailed feedback on the example sections etc., will do so in the second iteration (also, I can then help to formulate some specific text passages and highlight important things). For now, it is much more important that we have a compelling final figure that we can display, without having to compromise on MC resolution vs runtime duration.

Therefore, my single request in this first iteration: let's get the simulation scope straight. That is:
* Get rid of all those "recommended" numbers
* Set the settings to appropriate figures that **really** give a nice result. As a user, I would absolutely not be convinced by these low-MC figures. You can really use high settings for that.
* In the code, use these settings **by default** and dynamically switch to an absolute minimum setting (e.g. MC=1, iterations=1, etc) depending on an external `SMOKE_TEST` environment variable.

Once we are there, let's have a look at the figure together and decide if we are happy 👍🏼

**Comment by AVHopp** [comment]
Closed since we agreed upon a different kind of TL Backtesting example
